### PR TITLE
Use the right TLS model for CloudABI.

### PR DESCRIPTION
**Pull request by EdSchouten**
CloudABI doesn't do dynamic linking. For this reason, there is no need
to handle any other TLS model than local-exec. CloudABI's C library
doesn't provide a __tls_get_addr() function to do Dynamic TLS.

By forcing local-exec to be used here, we ensure that we don't generate
function calls to __tls_get_addr().

### CODE REVIEW

**Comment by kennytm** [comment]
I'm not familiar with TLS, but from what I searched, the `initial-exec` model also doesn't generate `__tls_get_addr()`. Could you explain why `local-exec` is chosen, besides "it doesn't do dynamic linking"? Thanks.

**Comment by EdSchouten** [comment]
Thanks for the review! It looks like `initial-exec` would work for CloudABI as well. That also doesn't emit any calls to `__tls_get_addr()`.

The main difference between `local-exec` and `initial-exec` is that the latter emits code that has one more level of indirection: it accesses the GOT (Global Offset Table) to obtain the effective address of a thread-local variable. Using a GOT is advantageous when doing dynamic linking, because it means that multiple accesses of the same variable within one library are folded into just one startup-time relocation.

With `local-exec`, the offset is expanded into every instruction that attempts to access it directly. This is optimal and for CloudABI, it is also sufficient.

**Comment by kennytm** [comment]
Thanks for explanation! Could you summarize those (https://github.com/rust-lang/rust/pull/47110#issue-285328124 + https://github.com/rust-lang/rust/pull/47110#issuecomment-355034069) as a code-comment? r=me after that.

**Comment by EdSchouten** [comment]
Sure thing!
